### PR TITLE
Sqs

### PR DESCRIPTION
**Pull request by McCarou**

### CODE REVIEW

File: adapter/event/sqs/adapter.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/sqs"
+	"github.com/radianteam/framework/adapter"
+)
+
+type SqsConfig struct {
+	Host            string `json:"host,omitempty" config:"host,required"`
+	Port            int16  `json:"port,omitempty" config:"port,required"`
+	Region          string `json:"region,omitempty" config:"region,required"`
+	AccessKeyId     string `json:"awsAccessKeyId,omitempty" config:"awsAccessKeyId,required"`
```
Comment (radianteam) [inline adapter/event/sqs/adapter.go:17]: all json and config params must be in snake case to save common style, like in other adapters

Comment (radianteam) [inline adapter/event/sqs/adapter.go:17]: Use capital letters in json and config params. Looks better :)

File: adapter/event/sqs/adapter.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+func NewSqsAdapter(name string, config *SqsConfig) *SqsAdapter {
+	return &SqsAdapter{BaseAdapter: adapter.NewBaseAdapter(name), config: config}
+}
+
+func (a *SqsAdapter) Setup() (err error) {
+	endpoint := fmt.Sprintf("http://%s:%d/", a.config.Host, a.config.Port)
+	cfg := aws.Config{
+		Region:      aws.String(endpoints.UsEast1RegionID),
```
Comment (radianteam) [inline adapter/event/sqs/adapter.go:39]: Hardcode. Make a config param for it

File: adapter/event/sqs/adapter.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	}
+
+	a.session = session.Must(session.NewSession(&cfg))
+	a.client = sqs.New(a.session)
+
+	return
+}
+
+func (a *SqsAdapter) Close() error {
+	return nil
```
Comment (radianteam) [inline adapter/event/sqs/adapter.go:51]: Closing sessions and clients?

File: adapter/event/sqs/adapter.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+func (a *SqsAdapter) CreateQueue(input *sqs.CreateQueueInput) (err error) {
+	_, err = a.client.CreateQueue(input)
```
Comment (radianteam) [inline adapter/event/sqs/adapter.go:55]: It is better to facade internal functions params with normal ones

File: example/sqs/README.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
+# Example: simple REST service
+
+1. [`Manual`](#1-manual) 
+2. [`Docker compose`](#2-docker-compose)
+
+## 1 Manual
+
+TODO
```
Comment (radianteam) [inline example/sqs/README.md:8]: Implement please

File: example/sqs/README.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
+# Example: simple REST service
```
Comment (radianteam) [inline example/sqs/README.md:1]: Copy-paste

File: example/sqs/docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    depends_on:
+      localstack:
+        condition: service_healthy
+  localstack:
+    container_name: "${LOCALSTACK_DOCKER_NAME-localstack_main}"
+    image: localstack/localstack
+    ports:
+      - "127.0.0.1:4566:4566"            # LocalStack Gateway
+      - "127.0.0.1:4510-4559:4510-4559"  # external services port range
+      - "127.0.0.1:53:53"                # DNS config (only required for Pro)
```
Comment (radianteam) [inline example/sqs/docker-compose.yml:22]: Delete unused ports and environment variables

File: example/sqs/main.go [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+	"strconv"
+)
+
+const (
+	inQueue    = "in"
+	outQueue   = "out"
+	sqsAdapter = "sqs-adapter"
+)
+
+func handlerRestIn(c *gin.Context, wc *worker.WorkerAdapters) {
```
Comment (radianteam) [inline example/sqs/main.go:23]: Add comments to explain how your example works. See another examples

File: worker/event/sqs/worker.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/sqs"
+	sqs_adapter "github.com/radianteam/framework/adapter/event/sqs"
+	"github.com/radianteam/framework/worker"
+	"sync"
+)
+
+type SqsWorkerRegFunc func(d *sqs.Message, wc *worker.WorkerAdapters) error
+
+type SqsConfig struct {
```
Comment (radianteam) [inline worker/event/sqs/worker.go:13]: AWS can have a shared file with configuration. Look at the S3 example. Make it compatible

File: worker/event/sqs/worker.go [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+	w.Logger.Info("Setting up Sqs Events")
+}
+
+func (w *SqsEventsWorker) Run() {
+	w.Logger.Info("Running Sqs Events Worker")
+
+	wg := sync.WaitGroup{}
+
+	adapterSqsConfig := &sqs_adapter.SqsConfig{
+		Host:            "localstack",
```
Comment (radianteam) [inline worker/event/sqs/worker.go:55]: HARDCODE!!!

File: worker/event/sqs/worker.go [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+	for {
+		if w.stopPolling {
+			break
+		}
+
+		for queueName, handler := range w.handlers {
+			wg.Add(1)
+
+			go func(qName string, handler SqsWorkerRegFunc) {
+				defer wg.Done()
```
Comment (radianteam) [inline worker/event/sqs/worker.go:81]: go routines inside eternal loop? Something wrong with this code

**Review by radianteam — COMMENTED** [review]
Please check comments

File: adapter/event/sqs/adapter.go (hunk 3 above)
Comment (McCarou) [reply adapter/event/sqs/adapter.go:55]: They are not required to be closed

File: worker/event/sqs/worker.go (hunk 9 above)
Comment (McCarou) [reply worker/event/sqs/worker.go:13]: done

File: adapter/event/sqs/adapter.go [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+	}
+
+	sendMessageInput := &sqs.SendMessageInput{QueueUrl: aws.String(queueUrl), MessageBody: aws.String(message)}
+	_, err = a.client.SendMessage(sendMessageInput)
+
+	return
+}
+
+func (a *SqsAdapter) Consume(queueUrl string) ([]*sqs.Message, error) {
+	receiveMessageInput := &sqs.ReceiveMessageInput{QueueUrl: aws.String(queueUrl)}
```
Comment (radianteam) [inline adapter/event/sqs/adapter.go:122]: Append WaitTimeSeconds

File: example/sqs/README.md [hunk 13]
```diff
@@ -0,0 +1,49 @@
+# Example: simple SQS service
+
+## Docker compose
```
Comment (radianteam) [inline example/sqs/README.md:3]: append manual guide like in other examples

File: example/sqs/docker-compose.yml [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+    container_name: "${LOCALSTACK_DOCKER_NAME-localstack_main}"
+    image: localstack/localstack
+    ports:
+      - "127.0.0.1:4566:4566"            # LocalStack Gateway
+      - "127.0.0.1:4510-4559:4510-4559"  # external services port range
+    environment:
+      - DEBUG=${DEBUG-}
+      - PERSISTENCE=${PERSISTENCE-}
+      - LAMBDA_EXECUTOR=${LAMBDA_EXECUTOR-}
+      - LOCALSTACK_API_KEY=${LOCALSTACK_API_KEY-}  # only required for Pro
```
Comment (radianteam) [inline example/sqs/docker-compose.yml:26]: Unnecessary

File: example/sqs/main.go [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+
+	// setup sqs adapter
+	adapterSqsConfig := &sqs_adapter.SqsConfig{
+		Host:            "localstack",
+		Port:            4566,
+		AccessKeyID:     "test_key_id",
+		SecretAccessKey: "test_secret_access_key",
+		SessionToken:    "test_token",
+		Region:          "us-east-2",
+		PollingTime:     10,
```
Comment (radianteam) [inline example/sqs/main.go:73]: Where is it used?

File: example/sqs/main.go [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+	}
+	workerSqs := sqs_worker.NewSqsEventsWorker("service_sqs", sqsWorkerConfig)
+	workerSqs.SetEvent(inQueue, fromInToOutQueueHandler)
+	workerSqs.SetAdapter(adapterSqs)
+
+	radian.AddWorker(workerSqs)
+	radian.AddWorker(workerRest)
+
+	// run the framework
+	radian.Run([]string{"service_rest", "service_sqs"})
```
Comment (radianteam) [inline example/sqs/main.go:107]: Use RunAll instead. It is better

File: worker/event/sqs/worker.go [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+	}
+	adapter := sqs_adapter.NewSqsAdapter("sqs-consumer", adapterSqsConfig)
+	err := adapter.Setup()
+	if err != nil {
+		w.Logger.Errorf("Failed to run '%s' worker during adapter configuration: %s", w.GetName(), err)
+
+		return
+	}
+
+	w.waitChan = make(chan bool)
```
Comment (radianteam) [inline worker/event/sqs/worker.go:73]: Make it after the loop

File: worker/event/sqs/worker.go [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+						return
+					}
+				}
+
+				w.Logger.Infof("Consuming queue '%s' stopped", qName)
+			}
+		}(queueName, handler)
+	}
+
+	wg.Wait()
```
Comment (radianteam) [inline worker/event/sqs/worker.go:128]: wait for the channel waitChan. Not only the wait group

File: worker/event/sqs/worker.go [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+				}
+
+				if err != nil {
+					w.Logger.Errorf("Failed to obtain url for queue '%s': %v", qName, err)
+
+					return
+				}
+
+				msgs, err := adapter.Consume(qName)
+				if err != nil {
```
Comment (radianteam) [inline worker/event/sqs/worker.go:94]: Check possibility to define a message limit

**Pull request merged**